### PR TITLE
[Feature] Debug logging and NVTX annotations

### LINKED ISSUE #166: [Feature] Debug logging and NVTX annotations
@awennersteen Could you please add a description ?

### PR DESCRIPTION
**Pull request by awennersteen**
Closes #166

### CODE REVIEW

File: pyqtorch/circuit.py [hunk 1]
```diff
@@ -1,15 +1,38 @@
...
 from torch.nn import Module, ModuleList
 
 from pyqtorch.utils import DiffMode, State, inner_prod, zero_state
 
 logger = getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+
+def forward_hook(*args, **kwargs) -> None:  # type: ignore[no-untyped-def]
+    logger.debug("Forward complete")
```
Comment (awennersteen) [inline pyqtorch/circuit.py:19]: The log appears to be too verbose for DEBUG, Might define a TRACE level

**Comment by dominikandreasseitz** [comment]
thanks @awennersteen , have you tested it in conjunction with qadence?

**Review by dominikandreasseitz — APPROVED** [review]
lgtm, would be nice to see how the logs actually look like for a standard usecase

File: pyproject.toml [hunk 2]
```diff
@@ -63,20 +63,20 @@ build = "mkdocs build --clean --strict"
...
-"__init__.py" = ["F401"]
+[tool.ruff.lint.per-file-ignores]
+"__init__.py" = ["F401", "E402"]
 
-[tool.ruff.mccabe]
+[tool.ruff.lint.mccabe]
 max-complexity = 15
 
-[tool.ruff.flake8-quotes]
+[tool.ruff.lint.flake8-quotes]
```
Comment (awennersteen) [inline pyproject.toml:79]: Requested by recent versions

File: pyqtorch/primitive.py [hunk 3]
```diff
@@ -19,6 +39,23 @@ def __init__(self, pauli: Tensor, target: int) -> None:
...
+            self.register_full_backward_pre_hook(pre_backward_hook)
+
+    def __key(self) -> tuple:
+        return self.qubit_support
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, type(self)):
+            return self.__key() == other.__key()
+        else:
+            return False
```
Comment (awennersteen) [inline pyqtorch/primitive.py:57]: I don't know where this came from

File: pyproject.toml (hunk 2 above)
Comment (dominikandreasseitz) [reply pyproject.toml:79]: got it

**Pull request merged**